### PR TITLE
JAX Docs tutorials: quickstart

### PR DESCRIPTION
**Pull request by jakevdp**
Part of #18585

### CODE REVIEW

File: docs/tutorials/quickstart.md [hunk 1]
```diff
@@ -1,8 +1,194 @@
...
-```{note}
-This is a placeholder for a section in the new {ref}`jax-tutorials`.
+**JAX is array-oriented computation (*a la* [NumPy](https://numpy.org/)), with automatic differentiation and JIT compilation for high-performance machine learning research**.
+
+This document gives a quick overview of essential JAX features:
+
+* JAX provides a unified NumPy-like interface to computations that run on CPU, GPU, or TPU, in single-host or distributed settings.
+* JAX features built-in Just-In-Time compilation via [Open XLA](https://github.com/openxla), an open-source machine learning compiler ecosystem.
+* JAX programs support efficient evaluation of gradients via its automatic differentiation transformations.
+* JAX programs can be automatically vectorized to efficiently map custom kernels across batches.
```
Comment (mattjj) [inline docs/tutorials/quickstart.md:24]:
```suggestion
* JAX functions can be automatically vectorized to efficiently map them over arrays representing batches of inputs.
```

File: docs/tutorials/quickstart.md [hunk 2]
```diff
@@ -1,8 +1,194 @@
...
+
+## Installation
+
+JAX can be installed for CPU on Linux, Windows, and OSX directly from the [Python Package Index](https://pypi.org/project/jax/):
+```
+pip install "jax[cpu]"
+```
+For more detailed installation information, including installation with GPU support, sett {ref}`installation`.
```
Comment (mattjj) [inline docs/tutorials/quickstart.md:32]:
```suggestion
For more detailed installation information, including installation with GPU support, see {ref}`installation`.
```

File: docs/tutorials/quickstart.md [hunk 3]
```diff
@@ -1,8 +1,194 @@
...
 # Quickstart
 
-```{note}
-This is a placeholder for a section in the new {ref}`jax-tutorials`.
+**JAX is array-oriented computation (*a la* [NumPy](https://numpy.org/)), with automatic differentiation and JIT compilation for high-performance machine learning research**.
+
+This document gives a quick overview of essential JAX features:
+
+* JAX provides a unified NumPy-like interface to computations that run on CPU, GPU, or TPU, in single-host or distributed settings.
+* JAX features built-in Just-In-Time compilation via [Open XLA](https://github.com/openxla), an open-source machine learning compiler ecosystem.
```
Comment (mattjj) [inline docs/tutorials/quickstart.md:22]:
```suggestion
* JAX features built-in Just-In-Time (JIT) compilation via [Open XLA](https://github.com/openxla), an open-source machine learning compiler ecosystem.
```

File: docs/tutorials/quickstart.md [hunk 4]
```diff
@@ -1,8 +1,194 @@
...
+
 # Quickstart
 
-```{note}
-This is a placeholder for a section in the new {ref}`jax-tutorials`.
+**JAX is array-oriented computation (*a la* [NumPy](https://numpy.org/)), with automatic differentiation and JIT compilation for high-performance machine learning research**.
+
+This document gives a quick overview of essential JAX features:
+
+* JAX provides a unified NumPy-like interface to computations that run on CPU, GPU, or TPU, in single-host or distributed settings.
```
Comment (mattjj) [inline docs/tutorials/quickstart.md:21]:
```suggestion
* JAX provides a unified NumPy-like interface to computations that run on CPU, GPU, or TPU, in local or distributed settings.
```

File: docs/tutorials/quickstart.md [hunk 5]
```diff
@@ -1,8 +1,194 @@
...
+  return lmbda * jnp.where(x > 0, x, alpha * jnp.exp(x) - alpha)
+
+x = jnp.arange(5.0)
+print(selu(x))
+```
+
+You'll find a few differences between JAX arrays and NumPy arrays once you begin digging-in;
+these are explored in  [🔪 JAX - The Sharp Bits 🔪](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html).
+
+## Using {func}`~jax.jit` to speed up functions
```
Comment (mattjj) [inline docs/tutorials/quickstart.md:57]: Maybe in this section's header, or maybe just at the end of its body, we should call out automatic parallelization over multiple devices. We don't want to bury the lede too much!

File: docs/tutorials/quickstart.md [hunk 6]
```diff
@@ -1,8 +1,194 @@
...
+print(hessian(sum_logistic)(x_small))
+```
+
+This kind of composition produces efficient code in practice; this is more-or-less how JAX's built-in {func}`jax.hessian` function is implemented.
+
+For more on automatic differentiation in JAX, see {ref}`automatic-differentiation`.
+
+## Auto-vectorization with {func}`~jax.vmap`
+
+JAX has one more transformation in its API that you might find useful: {func}`~jax.vmap`, the vectorizing map.
```
Comment (mattjj) [inline docs/tutorials/quickstart.md:140]:
```suggestion
Another useful transformation is {func}`~jax.vmap`, the vectorizing map.
```

**Review by mattjj — APPROVED** [review]
LGTM, though we should iterate more on this. Re-reading it now after all this time, I realize we have better examples (e.g. in slide decks), and we might choose different phrasings.

File: docs/tutorials/quickstart.md [hunk 7]
```diff
@@ -1,8 +1,194 @@
...
+  display_name: Python 3
+  language: python
+  name: python3
+---
+
 # Quickstart
 
-```{note}
-This is a placeholder for a section in the new {ref}`jax-tutorials`.
+**JAX is a library for array-oriented numerical computation (*a la* [NumPy](https://numpy.org/)), with automatic differentiation and JIT compilation to enable high-performance machine learning research**.
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:17]: Nit: Use more accessible language for readers who don't know this French expression:

```suggestion
**JAX is a library for array-oriented numerical computation (like [NumPy](https://numpy.org/)), with automatic differentiation and JIT compilation to enable high-performance machine learning research**.
```

File: docs/tutorials/quickstart.md [hunk 8]
```diff
@@ -1,8 +1,194 @@
...
+
+This document gives a quick overview of essential JAX features:
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:19]: Nit:

```suggestion
This document provides a quick overview of essential JAX features, so you can get started with JAX quickly:
```

WDYT>

File: docs/tutorials/quickstart.md [hunk 9]
```diff
@@ -1,8 +1,194 @@
...
+
+* JAX provides a unified NumPy-like interface to computations that run on CPU, GPU, or TPU, in local or distributed settings.
+* JAX features built-in Just-In-Time (JIT) compilation via [Open XLA](https://github.com/openxla), an open-source machine learning compiler ecosystem.
+* JAX programs support efficient evaluation of gradients via its automatic differentiation transformations.
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:23]:
```suggestion
* JAX functions support efficient evaluation of gradients via its automatic differentiation transformations.
```

@mattjj may have proposed a similar change elsewhere in the doc (JAX programs -> JAX functions)

File: docs/tutorials/quickstart.md [hunk 10]
```diff
@@ -1,8 +1,194 @@
...
+* JAX functions can be automatically vectorized to efficiently map them over arrays representing batches of inputs.
+
+## Installation
+
+JAX can be installed for CPU on Linux, Windows, and OSX directly from the [Python Package Index](https://pypi.org/project/jax/):
+```
+pip install "jax[cpu]"
+```
+For more detailed installation information, including installation with GPU support, see {ref}`installation`.
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:32]: Nit: Use more accessibility-friendly language, as not every reader can "see" (they can "read", for example).

```suggestion
For more detailed installation information, including installation with GPU and TPU support, refer to {ref}`installation`.
```

File: docs/tutorials/quickstart.md [hunk 11]
```diff
@@ -1,8 +1,194 @@
...
+This document gives a quick overview of essential JAX features:
+
+* JAX provides a unified NumPy-like interface to computations that run on CPU, GPU, or TPU, in local or distributed settings.
+* JAX features built-in Just-In-Time (JIT) compilation via [Open XLA](https://github.com/openxla), an open-source machine learning compiler ecosystem.
+* JAX programs support efficient evaluation of gradients via its automatic differentiation transformations.
+* JAX functions can be automatically vectorized to efficiently map them over arrays representing batches of inputs.
+
+## Installation
+
+JAX can be installed for CPU on Linux, Windows, and OSX directly from the [Python Package Index](https://pypi.org/project/jax/):
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:28]:
```suggestion
JAX can be installed for CPU on Linux, Windows, and macOS directly from the [Python Package Index](https://pypi.org/project/jax/):
```

File: docs/tutorials/quickstart.md [hunk 12]
```diff
@@ -1,8 +1,194 @@
...
+```
+
+You'll find a few differences between JAX arrays and NumPy arrays once you begin digging-in;
+these are explored in  [🔪 JAX - The Sharp Bits 🔪](https://jax.readthedocs.io/en/latest/notebooks/Common_Gotchas_in_JAX.html).
+
+## Using {func}`~jax.jit` to speed up functions
+JAX runs transparently on the GPU or TPU (falling back to CPU if you don't have one). However, in the above example, JAX is dispatching kernels to the chip one operation at a time. If we have a sequence of operations, we can use the `@jit` decorator to compile multiple operations together using XLA.
+
+We can use IPython's `%timeit` to quickly benchmark our `selu` function, using `block_until_ready()` to
+account for JAX's dynamic dispatch (See {ref}`async-dispatch`):
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:61]: Nit: Use more accessibility-friendly language.

```suggestion
account for JAX's dynamic dispatch (refer to {ref}`async-dispatch`):
```

File: docs/tutorials/quickstart.md [hunk 13]
```diff
@@ -1,8 +1,194 @@
...
+from jax import random
+
+key = random.key(1701)
+x = random.normal(key, (1_000_000,))
+%timeit selu(x).block_until_ready()
+```
+
+(notice we've used {mod}`jax.random` to generate some random numbers; for details on how to generate random
+numbers in JAX, see {ref}`pseudorandom-numbers`).
+
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:73]: Nit:

```suggestion
Notice we used {mod}`jax.random` to generate some random numbers. For details on how to generate random
numbers in JAX, check out {ref}`pseudorandom-numbers`).

```

File: docs/tutorials/quickstart.md [hunk 14]
```diff
@@ -1,8 +1,194 @@
...
+We can speed the execution of this function with `@jit`, which will jit-compile the first time `selu` is called and will be cached thereafter.
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:74]:
```suggestion
We can speed up the execution of this function with `@jit`, which will JIT-compile the first time `selu` is called and be cached thereafter.
```

File: docs/tutorials/quickstart.md [hunk 15]
```diff
@@ -1,8 +1,194 @@
...
+
+selu_jit = jit(selu)
+_ = selu_jit(x)  # compiles on first call
+%timeit selu_jit(x).block_until_ready()
+```
+
+The above timing represent execution on CPU, but the same code can be run on GPU or TPU for
+an even greater speedup.
+
+For more on JIT compilation in JAX, see {ref}`jit-compilation`.
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:87]: it: Use more accessibility-friendly language.

```suggestion
For more on JIT compilation in JAX, check out {ref}`jit-compilation`.
```

File: docs/tutorials/quickstart.md [hunk 16]
```diff
@@ -1,8 +1,194 @@
...
+
+## Taking derivatives with {func}`~jax.grad`
+
+In addition to evaluating numerical functions, we also want to transform them.
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:91]: Maybe:

```suggestion
In addition to evaluating numerical functions, we can also to transform them.
```

File: docs/tutorials/quickstart.md [hunk 17]
```diff
@@ -1,8 +1,194 @@
...
+One transformation is [automatic differentiation](https://en.wikipedia.org/wiki/Automatic_differentiation).
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:92]:
```suggestion
One transformation is [automatic differentiation (autodiff)](https://en.wikipedia.org/wiki/Automatic_differentiation).
```

Since we're using the shortened version later.

File: docs/tutorials/quickstart.md [hunk 18]
```diff
@@ -1,8 +1,194 @@
...
+print(hessian(sum_logistic)(x_small))
+```
+
+This kind of composition produces efficient code in practice; this is more-or-less how JAX's built-in {func}`jax.hessian` function is implemented.
+
+For more on automatic differentiation in JAX, see {ref}`automatic-differentiation`.
+
+## Auto-vectorization with {func}`~jax.vmap`
+
+Another useful transformation is {func}`~jax.vmap`, the vectorizing map.
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:140]: Nit

```suggestion
Another useful JAX transformation is {func}`~jax.vmap`, the vectorizing map.
```

File: docs/tutorials/quickstart.md [hunk 19]
```diff
@@ -1,8 +1,194 @@
...
+def vmap_batched_apply_matrix(v_batched):
+  return vmap(apply_matrix)(v_batched)
+
+print('Auto-vectorized with vmap')
+%timeit vmap_batched_apply_matrix(batched_x).block_until_ready()
+```
+
+Of course, {func}`~jax.vmap` can be arbitrarily composed with {func}`~jax.jit`, {func}`~jax.grad`, and any other JAX transformation.
+
+For more on automatic vectorization in JAX, see {ref}`automatic-vectorization`.
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:192]: Nit: Use more accessibility-friendly language.

```suggestion
For more on automatic vectorization in JAX, check {ref}`automatic-vectorization`.
```

File: docs/tutorials/quickstart.md [hunk 20]
```diff
@@ -1,8 +1,194 @@
...
+
+This is just a taste of what JAX can do. We're really excited to see what you do with it!
```
Comment (8bitmp3) [inline docs/tutorials/quickstart.md:194]: Maybe we could add the next steps (each of the {transform name} 101s).

```
## Next steps

Discover more about each of the transforms you learned about in this Quickstart by...

- ... 101
- ... 101
- ... 101
```

**Comment by jakevdp** [comment]
Thanks for the reviews! Let's plan to continue iterating on this: in particular, I think as we work on the next section (how to think in JAX) we'll want to de-dupe some of the content into quickstart, so we may have a chance to take a look at this again.

**Comment by 8bitmp3** [comment]
:rocket:  @jakevdp  Thank you!
